### ISSUE TITLE
Discussion on Feature request: Document Unit specification

### ISSUE THREAD

**Issue by trenta3** [HelloZeroNet/ZeroNet#1513]
# Proposed Feature
Add a way to specify what the document unit (i.e. informations that are in a single page) for each site is. To explain it better I'm going to use ZeroTalk as an example, but first:
 
## Why do we want such a thing?
* For a possible search engine: to be able to index documents by document unit for each site (this would allow not to scrape html pages, be able to only index actual content rather than fixed header/footer/menu and allow it to be very fast, since it would only need to do some queries to the database)
* For possibly archiving sites: for example after a month that a topic has no more replies (the document unit ids do not change) simply move those "lines" to another site (impossible to do at current state)

## Example of proposed `docunit.json` for ZeroTalk
```
{
    "unit_topic": {
		"url": "?Topic:<topic_id:int>",
                "id_list": "SELECT topic_id FROM topic",
		"db_lines": {
			"topic": "SELECT topic_id FROM topic WHERE topic_id = :topic_id",
			"comment": "SELECT comment_id FROM comment WHERE topic_uri = :topic_id",
			"comment_vote": "SELECT ...",
			"topic_vote": "SELECT ..."
		}
    }
}
```
### Explanation of the above stub
The global object has one key for each "type of document unit" (ZeroTalk has only one kind, i.e. topics, but a size could have many). For each of these we have three compulsory keys:
* `url`: That gives the url (relative to the site one) to get to view that document unit (useful for search engines, but also can be used in reverse: when we are visiting a page and want, as an example, to locally save the content in it for archival purposes, we could parse the url we are in and use the parameters in it to get the important lines in each of the database table).
  The parameters are in the form `<param_name:param_type>`: param_type is useful for parsing, and param_name for referring at it, both in the `id_list` and in `db_lines`.
* `id_list`: Database query to select the valid values for parameters (i.e. to get all of the topics the site currently has). The returned table must return under the column `param_name` the possible `param_value`s.
* `db_lines`: This is an object, where the keys are the database tables and the values are the database queries that return the ids identifying the objects pertaining to that document unit.

I'm not so good at explaining things, as you can see by reading the above stuff. I hope it is clear enough. If not, do not hesitate to ask.

#### Possible proposed parameters for url
* `<integer:int>`: `/[0-9]+/`
* `<string:str>`: `/[a-zA-Z0-9\_\-]+/` (may be changed to account for alphanumeric characters in other languages, diacritics, ...)
* `<everything:all>`: Matches the whole url parameter

## Pro and Cons
**Pro**:
* No need to write code to have it done, just need to agree on a format for it, so *please add your two cents* and propose improvements for the format.

**Comment by HelloZeroNet** [comment]
Yes, something like this will be required for efficient site archiving/checkpoints.

For searching we have similar in dbschema.json, for example for ZeroTalk:
```
	"feeds": {
		"Topics": "SELECT title AS title, body AS body, added AS date_added, 'topic' AS type, '?Topic:' || topic.topic_id || '_' || topic_creator_json.directory AS url FROM topic LEFT JOIN json AS topic_creator_json ON (topic_creator_json.json_id = topic.json_id)",
		"Comments": "SELECT 'comment' AS type, comment.added AS date_added, topic.title AS title, commenter_user.value || ': ' || comment.body AS body, topic_creator_json.directory AS topic_creator_address, topic.topic_id || '_' || topic_creator_json.directory AS row_topic_uri, '?Topic:' || topic.topic_id || '_' || topic_creator_json.directory AS url FROM topic LEFT JOIN json AS topic_creator_json ON (topic_creator_json.json_id = topic.json_id) LEFT JOIN comment ON (comment.topic_uri = row_topic_uri) LEFT JOIN json AS commenter_json ON (commenter_json.json_id = comment.json_id) LEFT JOIN json AS commenter_content ON (commenter_content.directory = commenter_json.directory AND commenter_content.file_name = 'content.json') LEFT JOIN keyvalue AS commenter_user ON (commenter_user.json_id = commenter_content.json_id AND commenter_user.key = 'cert_user_id')"
	}
```

**Comment by trenta3** [comment]
The one thing I don't get is how can a peer prove to others that the document unit he is sending them is legit: I mean usually each peer has only one file with all its data inside and it only signs the file as a whole.

Maybe we should change the signature to be the signature of a sort of merkle hash of the json document. In such a way one can, given the original signature and json document, prove to another party that the piece of information is legit by giving them the hashes from The root to the interested leaf.

If someone has better ideas on how to do that, i'd like to hear them.
